### PR TITLE
Delete ImmutableHashMap

### PR DESCRIPTION
**Pull request by jasonmalinowski**
This doesn't seem to be notably faster than just using the framework library; in a case where we're doing an excessive number of GetProject calls this only is saving us a few dozen milliseconds over all of loading Roslyn.sln.

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
For those following along at home, I'm not going to merge this until I can do an RPS run internally, and right now we're still sorting things out internally.

**Comment by jinujoseph** [comment]
pre-approved ,  conditional on what we find on RPS run

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski can this be moved forward with?

**Comment by sharwell** [comment]
> This doesn't seem to be notably faster than just using the framework library

This is because both implementations use the same inefficient storage strategy. 😄 I'm working on improving this space but this PR wouldn't block that work.

**Comment by sharwell** [comment]
`ImmutableSegmentedDictionary<TKey, TValue>` is now available for use in Roslyn if you need an immutable dictionary with high-performance indexing.

**Comment by jasonmalinowski** [comment]
@sharwell It's been so long I don't even remember how "slow" the standard one is in this case....IIRC there was little to no cost with the standard one but will look again.

File: src/Workspaces/Core/Portable/Workspace/Solution/Project.cs [hunk 1]
```diff
@@ -227,19 +226,40 @@ public bool ContainsAnalyzerConfigDocument(DocumentId documentId)
         /// Get the document in this project with the specified document Id.
         /// </summary>
         public Document? GetDocument(DocumentId documentId)
-            => ImmutableHashMapExtensions.GetOrAdd(ref _idToDocumentMap, documentId, s_tryCreateDocumentFunction, this);
+        {
+            if (ContainsDocument(documentId))
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/Project.cs:230]: ImmutableHashMapExtensions.GetORAdd had an important behavior that meant if the lookup into the dictionary missed, it'd call the creation function, and if it returned null would then return null. The standard ImmutableInterlocked.GetOrAdd does not do such a thing; this pattern is the pattern we use in Solution.GetProject so I'm using it here until we have data to argue for something better/different.

File: src/Workspaces/Core/Portable/Workspace/Solution/Project.cs [hunk 2]
```diff
@@ -407,14 +427,14 @@ static bool FilterMatches(DeclaredSymbolInfo info, SymbolFilter filter)
             return results.Any(b => b);
         }
 
-        private static readonly Func<DocumentId, Project, Document?> s_tryCreateDocumentFunction =
-            (documentId, project) => project._projectState.DocumentStates.TryGetState(documentId, out var state) ? new Document(project, state) : null;
+        private static readonly Func<DocumentId, Project, Document> s_tryCreateDocumentFunction =
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Workspace/Solution/Project.cs:430]: Applies to all `s_try*` which are no longer conditional

```suggestion
        private static readonly Func<DocumentId, Project, Document> s_createDocumentFunction =
```

**Review by sharwell — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski can we move forward with this?

**Comment by jasonmalinowski** [comment]
Obsoleted by #72520.

**Pull request closed without merging**